File: src/api/journal.ts
/**
 * Journal API Service
 */
import { apiClient, PaginatedResponse } from './client';

export interface JournalEntry {
  id: string;
  user_id: string;
  title: string;
  content: string;
  date: string;
  cover_image?: string;
  images: string[];
  mood?: string;
  tags: string[];
  is_private: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreateJournalEntryRequest {
  title: string;
  content: string;
  date?: string;
  cover_image?: string;
  images?: string[];
  mood?: string;
  tags?: string[];
  is_private?: boolean;
}

export interface UpdateJournalEntryRequest {
  title?: string;
  content?: string;
  cover_image?: string;
  images?: string[];
  mood?: string;
  tags?: string[];
  is_private?: boolean;
}

export const journalApi = {
  /**
   * Get journal entries with pagination
   */
  getEntries: async (params?: {
    start_date?: string;
    end_date?: string;
    page?: number;
    limit?: number;
  }): Promise<PaginatedResponse<JournalEntry>> => {
    const response = await apiClient.get<PaginatedResponse<JournalEntry>>('/journal/entries', {
      params,
    });
    return response.data;
  },

  /**
   * Get journal entry by ID
   */
  getEntry: async (id: string): Promise<JournalEntry> => {
    const response = await apiClient.get<JournalEntry>(`/journal/entries/${id}`);
    return response.data;
  },

  /**
   * Create journal entry
   */
  createEntry: async (data: CreateJournalEntryRequest): Promise<JournalEntry> => {
    const response = await apiClient.post<JournalEntry>('/journal/entries', data);
    return response.data;
  },

  /**
   * Update journal entry
   */
  updateEntry: async (id: string, data: UpdateJournalEntryRequest): Promise<JournalEntry> => {
    const response = await apiClient.put<JournalEntry>(`/journal/entries/${id}`, data);
    return response.data;
  },

  /**
   * Delete journal entry
   */
  deleteEntry: async (id: string): Promise<void> => {
    await apiClient.delete(`/journal/entries/${id}`);
  },
};

File: backend/routers/journal.py
"""
Journal router - Journaling endpoints
"""
from fastapi import APIRouter, Depends, Query
from datetime import datetime
from typing import Optional
from uuid import UUID
import math

from schemas.journal import (
    CreateJournalEntryRequest,
    UpdateJournalEntryRequest,
    JournalEntryResponse
)
from services.journal_service import journal_service
from core.auth import get_current_user, CurrentUser
from core.responses import Response, PaginatedResponse

router = APIRouter()


@router.post("/entries", response_model=Response[JournalEntryResponse])
async def create_journal_entry(
    request: CreateJournalEntryRequest,
    current_user: CurrentUser = Depends(get_current_user)
):
    """Create a new journal entry"""
    entry = await journal_service.create_entry(current_user.user_id, request)
    return Response(
        success=True,
        message="Journal entry created successfully",
        data=entry
    )


@router.get("/entries", response_model=PaginatedResponse[JournalEntryResponse])
async def get_journal_entries(
    start_date: Optional[datetime] = Query(None, description="Filter entries from this date"),
    end_date: Optional[datetime] = Query(None, description="Filter entries until this date"),
    page: int = Query(1, ge=1, description="Page number"),
    limit: int = Query(10, ge=1, le=50, description="Items per page"),
    current_user: CurrentUser = Depends(get_current_user)
):
    """Get journal entries with pagination"""
    entries, total = await journal_service.get_entries(
        user_id=current_user.user_id,
        start_date=start_date,
        end_date=end_date,
        page=page,
        limit=limit
    )
    
    return PaginatedResponse(
        success=True,
        message="Journal entries retrieved successfully",
        data=entries,
        page=page,
        limit=limit,
        total=total,
        total_pages=math.ceil(total / limit)
    )


@router.get("/entries/{entry_id}", response_model=Response[JournalEntryResponse])
async def get_journal_entry(
    entry_id: UUID,
    current_user: CurrentUser = Depends(get_current_user)
):
    """Get journal entry by ID"""
    entry = await journal_service.get_entry_by_id(current_user.user_id, entry_id)
    return Response(
        success=True,
        message="Journal entry retrieved successfully",
        data=entry
    )


@router.put("/entries/{entry_id}", response_model=Response[JournalEntryResponse])
async def update_journal_entry(
    entry_id: UUID,
    request: UpdateJournalEntryRequest,
    current_user: CurrentUser = Depends(get_current_user)
):
    """Update journal entry"""
    entry = await journal_service.update_entry(current_user.user_id, entry_id, request)
    return Response(
        success=True,
        message="Journal entry updated successfully",
        data=entry
    )


@router.delete("/entries/{entry_id}", response_model=Response[dict])
async def delete_journal_entry(
    entry_id: UUID,
    current_user: CurrentUser = Depends(get_current_user)
):
    """Delete journal entry"""
    await journal_service.delete_entry(current_user.user_id, entry_id)
    return Response(
        success=True,
        message="Journal entry deleted successfully",
        data={}
    )
